src/tools/cephfs_mirror/SnapReplicator.cc

```diff
@@ -0,0 +1,434 @@
+#include <dirent.h>
+#include <string.h>
+#include <limits.h>
+
+#include "SnapReplicator.h"
+
+using namespace std::literals::string_literals;
+
+bool operator ==(const struct timespec& lhs, const struct timespec& rhs);
+bool operator >(const struct timespec& lhs, const struct timespec& rhs);
+bool operator >=(const struct timespec& lhs, const struct timespec& rhs);
+
+
+SnapReplicator::SnapReplicator(struct ceph_mount_info *_src_mnt,
+                               struct ceph_mount_info *_dst_mnt,
+                               const std::string& _src_dir,
+                               const std::pair<std::string, std::string>& snaps)
+  : m_src_mnt(_src_mnt),
+    m_dst_mnt(_dst_mnt),
+    m_src_dir(_src_dir),
+    m_old_snap(snaps.first),
+    m_new_snap(snaps.second)
+{
+  char buf[128] = {0,};
+  if (ceph_conf_get(m_src_mnt, "client_snapdir", buf, sizeof(buf)) < 0)
+    std::clog << __func__ << ":" << __LINE__ << ": unable to get src client_snapdir" << std::endl;
+  else
+    m_src_snap_dir = buf;
+  if (ceph_conf_get(m_dst_mnt, "client_snapdir", buf, sizeof(buf)) < 0)
+    std::clog << __func__ << ":" << __LINE__ << ": unable to get dst client_snapdir" << std::endl;
+  else
+    m_dst_snap_dir = buf;
+}
+
+int SnapReplicator::handle_old_entry(const std::string& dir_name,
+                                     const std::string& old_dentry,
+                                     const struct dirent& de,
+                                     const struct ceph_statx& old_stx,
+                                     const struct ceph_statx& new_stx,
+                                     const struct timespec& old_rctime,
+                                     std::queue<std::string>& dir_queue)
+{
+  if (new_stx.stx_mtime > old_stx.stx_mtime or new_stx.stx_ctime > old_stx.stx_ctime) {
+    if (S_ISDIR(new_stx.stx_mode)) {
+      if (get_rctime(m_new_snap, "." + dir_name + "/"s + de.d_name) >= old_rctime)
+        dir_queue.push(dir_name + "/"s + de.d_name);
+    } else if (S_ISREG(new_stx.stx_mode)) {
+      if (copy_blocks(dir_name + "/"s + de.d_name, file_type_t::OLD_FILE, old_stx, new_stx) < 0) {
```

Comment:
Is there an advantage of copying the blocks and then truncating?

Reply:
not really ... no advantage of truncating **after** copying blocks
the ordered operations (copy blocks, truncate) and (truncate, copy blocks) should be idempotent
however, **copy blocks** is not an atomic operation; so the operations are idempotent only if they complete successfully